server/src/main/java/org/apache/iotdb/db/qp/strategy/optimizer/ConcatPathOptimizer.java

```diff
@@ -47,202 +47,88 @@
-          List<PartialPath> afterConcatUdfPaths = new ArrayList<>();
-          for (PartialPath fromPath : fromPaths) {
-            afterConcatUdfPaths.add(fromPath.concatPath(originUdfSuffixPath));
-          }
-          afterConcatUdfPathsList.add(afterConcatUdfPaths);
-        }
-        List<List<PartialPath>> extendedAfterConcatUdfPathsList = new ArrayList<>();
-        cartesianProduct(
-            afterConcatUdfPathsList, extendedAfterConcatUdfPathsList, 0, new ArrayList<>());
-
-        for (List<PartialPath> afterConcatUdfPaths : extendedAfterConcatUdfPathsList) {
-          afterConcatPaths.add(null);
-          extendListSafely(originAggregations, i, afterConcatAggregations);
-
-          afterConcatUdfList.add(
-              new UDFContext(originUdf.getName(), originUdf.getAttributes(), afterConcatUdfPaths));
-        }
-      } else { // non-udf
-        for (PartialPath fromPath : fromPaths) {
-          PartialPath fullPath = fromPath.concatPath(selectPath);
-          if (selectPath.isTsAliasExists()) {
-            fullPath.setTsAlias(selectPath.getTsAlias());
-          }
-          afterConcatPaths.add(fullPath);
-          extendListSafely(originAggregations, i, afterConcatAggregations);
+    if (queryOperator.isAlignByDevice() && !queryOperator.isLastQuery()
+        || queryOperator.getIndexType() != null) {
+      return;
+    }
 
-          afterConcatUdfList.add(null);
-        }
+    WildcardsRemover wildcardsRemover = new WildcardsRemover(this, queryOperator, fetchSize);
+    List<ResultColumn> resultColumns = new ArrayList<>();
+    for (ResultColumn resultColumn : queryOperator.getSelectOperator().getResultColumns()) {
+      resultColumn.removeWildcards(wildcardsRemover, resultColumns);
+      if (wildcardsRemover.checkIfPathNumberIsOverLimit(resultColumns)) {
+        break;
       }
     }
+    wildcardsRemover.checkIfSoffsetIsExceeded(resultColumns);
+    queryOperator.getSelectOperator().setResultColumns(resultColumns);
+  }
 
-    if (needRemoveStar) {
-      removeStarsInPath(
-          afterConcatPaths,
-          afterConcatAggregations,
-          afterConcatUdfList,
-          selectOperator,
-          limit,
-          offset,
-          maxDeduplicatedPathNum);
-    } else {
-      selectOperator.setSuffixPathList(afterConcatPaths);
+  private void concatFilter(QueryOperator queryOperator) throws LogicalOptimizeException {
+    FilterOperator filter = queryOperator.getFilterOperator();
+    if (filter == null) {
+      return;
     }
+
+    Set<PartialPath> filterPaths = new HashSet<>();
+    if (!queryOperator.isAlignByDevice() || queryOperator.isLastQuery()) {
+      // GROUP_BY_DEVICE leaves the concatFilter to PhysicalGenerator to optimize filter without
+      // prefix first
+      queryOperator.setFilterOperator(
+          concatFilter(queryOperator.getFromOperator().getPrefixPaths(), filter, filterPaths));
+    }
+    queryOperator.getFilterOperator().setPathSet(filterPaths);
```

Comment:
The logic is confusing here. Actually it's the same as `concatSelect`. If `alignbydevice`, return, otherwise concat filter.